cardano-crypto-class/src/Cardano/Crypto/SECP256K1/C.hs

```diff
@@ -1,78 +1,186 @@
 {-# LANGUAGE CApiFFI #-}
 {-# LANGUAGE DerivingVia #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds #-}
 
 module Cardano.Crypto.SECP256K1.C (
   SECP256k1Context,
   secpContextSignVerify,
   SECP256k1SchnorrExtraParams,
   secpContextCreate,
-  secpContextDestroy,
   secpKeyPairCreate,
   secpSchnorrSigSignCustom,
   secpKeyPairXOnlyPub,
   secpSchnorrSigVerify,
+  secpXOnlyPubkeySerialize,
+  secpXOnlyPubkeyParse,
+  secpCtxPtr,
+  secpEcPubkeyCreate,
+  secpEcdsaSign,
+  secpEcdsaVerify,
+  secpEcCompressed,
+  secpEcPubkeySerialize,
+  secpEcdsaSignatureSerializeCompact,
+  secpEcdsaSignatureParseCompact,
+  secpEcPubkeyParse,
   ) where
 
+import Control.Exception (mask_)
 import Data.Bits ((.|.))
+import Foreign.ForeignPtr (ForeignPtr, FinalizerPtr, newForeignPtr)
 import Foreign.Ptr (Ptr)
-import Foreign.C.Types (CUChar, CSize (CSize), CInt (CInt))
-import Cardano.Foreign
-import Cardano.Crypto.SECP256K1.Constants
+import System.IO.Unsafe (unsafePerformIO)
+import Foreign.C.Types (CUChar, CSize (CSize), CInt (CInt), CUInt (CUInt))
+import Cardano.Foreign (SizedPtr (SizedPtr))
+import Cardano.Crypto.SECP256K1.Constants (
+  SECP256K1_SCHNORR_KEYPAIR_BYTES,
+  SECP256K1_SCHNORR_PRIVKEY_BYTES,
+  SECP256K1_SCHNORR_SIGNATURE_BYTES,
+  SECP256K1_SCHNORR_PUBKEY_BYTES_INTERNAL,
+  SECP256K1_SCHNORR_PUBKEY_BYTES,
+  SECP256K1_ECDSA_PUBKEY_BYTES_INTERNAL,
+  SECP256K1_ECDSA_PRIVKEY_BYTES,
+  SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL,
+  SECP256K1_ECDSA_SIGNATURE_BYTES,
+  SECP256K1_ECDSA_MESSAGE_BYTES,
+  )
 
 data SECP256k1Context
 
 data SECP256k1SchnorrExtraParams
 
-foreign import capi "secp256k1.h secp256k1_context_create"
+-- We create a single context for both signing and verifying, which we use
+-- everywhere. This saves considerable time, and is safe, provided nobody
+-- outside cardano-base gets to touch it.
+--
+-- We do _not_ make this dupable, as the whole point is _not_ to compute it more
+-- than once!
+{-# NOINLINE secpCtxPtr #-}
+secpCtxPtr :: ForeignPtr SECP256k1Context
+secpCtxPtr = unsafePerformIO . mask_ $ do
+  ctx <- secpContextCreate secpContextSignVerify
```

Comment:
In practice (at least in the context of plutus scripts) we only need the verify context, not the sign verify. Would it be possible to globally initialise only the verify context (i.e. `secp256k1_context_create(SECP256K1_CONTEXT_VERIFY)`), and use the `sign | verify` only for testing?

Reply:
I mostly created a `sign | verify` context because the rest of the API needs it. There's no reason why I couldn't have two contexts and use the 'least permissive' one where needed. What's the advantage of limiting your context's capabilities like this?

Reply:
I don't think it is worth it. The only upside for doing this would be a very tiny reduction in amount of memory use and a slightly faster initialization which is done once at start up. The cost of much more complicated code for this optimization is way too high in my opinion.

My opinion, it is definitely not worth it.

Reply:
If this is the only real benefit, I agree. That's not significant enough, especially for something nobody but us ever interacts with.

Reply:
If the complexity of supporting such a change is high, then I agree that we should not do so.